libraries/rush-lib/src/utilities/templateUtilities.ts

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+// Matches a well-formed BEGIN macro starting a block section.
+// Example:  /*[BEGIN "DEMO"]*/
+//
+// Group #1 is the indentation spaces before the macro
+
+import { FileSystem, InternalError, NewlineKind } from '@rushstack/node-core-library';
+import { Colorize } from '@rushstack/terminal';
+import { Rush } from '../api/Rush';
+
+// Group #2 is the section name
+const BEGIN_MARCO_REGEXP: RegExp = /^(\s*)\/\*\[BEGIN "([A-Z]+)"\]\s*\*\/\s*$/;
+
+// Matches a well-formed END macro ending a block section.
+// Example:  /*[END "DEMO"]*/
+//
+// Group #1 is the indentation spaces before the macro
+// Group #2 is the section name
+const END_MACRO_REGEXP: RegExp = /^(\s*)\/\*\[END "([A-Z]+)"\]\s*\*\/\s*$/;
+
+// Matches a well-formed single-line section, including the space character after it
+// if present.
+// Example:  /*[LINE "HYPOTHETICAL"]*/
+//
+// Group #1 is the section name
+const LINE_MACRO_REGEXP: RegExp = /\/\*\[LINE "([A-Z]+)"\]\s*\*\/\s?/;
+
+// Matches a variable expansion.
+// Example:  [%RUSH_VERSION%]
+//
+// Group #1 is the variable name including the dollar sign
+const VARIABLE_MACRO_REGEXP: RegExp = /\[(%[A-Z0-9_]+%)\]/;
+
+// Matches anything that starts with "/*[" and ends with "]*/"
+// Used to catch malformed macro expressions
+const ANY_MACRO_REGEXP: RegExp = /\/\*\s*\[.*\]\s*\*\//;
+
+export class TemplateUtilities {
+  // Copy the template from sourcePath, transform any macros, and write the output to destinationPath.
+  //
+  // We implement a simple template engine.  "Single-line section" macros have this form:
+  //
+  //     /*[LINE "NAME"]*/ (content goes here)
+  //
+  // ...and when commented out will look like this:
+  //
+  //     // (content goes here)
+  //
+  // "Block section" macros have this form:
+  //
+  //     /*[BEGIN "NAME"]*/
+  //     (content goes
+  //     here)
+  //     /*[END "NAME"]*/
+  //
+  // ...and when commented out will look like this:
+  //
+  //     // (content goes
+  //     // here)
+  //
+  // Lastly, a variable expansion has this form:
+  //
+  //     // The value is [%NAME%].
+  //
+  // ...and when expanded with e.g. "123" will look like this:
+  //
+  //     // The value is 123.
+  //
+  // The section names must be one of the predefined names used by "rush init".
+  // A single-line section may appear inside a block section, in which case it will get
+  // commented twice.
+  public static copyTemplateFile(
+    sourcePath: string,
+    destinationPath: string,
+    overwrite: boolean,
+    demo: boolean = false
```

Comment:
Can you make this an options object instead?

Reply:
Changed